man/retro.1
.Dd August 2026
.Dt RETRO 1
.Os
.Sh RETRO
.Nm retro
.Nd "a modern, pragmatic forth development system"
.Sh SYNOPSIS
.Nm
.Op Fl h
.Op Fl i
.Op Fl t Ar filename
.Op Fl f Ar filename
.Op Fl p Ar filename
.Op Fl u Ar filename
.Op Fl r Ar filename
.Op Fl cs Ar delimiter
.Op Fl ce Ar delimiter
.Op Fl ts Ar delimiter
.Op Fl te Ar delimiter
.Op Ar filename Ar script-args
.Sh DESCRIPTION
RETRO is a modern, pragmatic Forth drawing influences from many sources.
It's clean, elegant, tiny, and easy to grasp and adapt to various uses.
.Pp
.Nm
is the main interface for interacting with Retro. It provides both
an interactive and a scripting model.
.Sh OPTIONS
.Bl -tag -width -indent
.It Fl h
Display a help screen.
.It Fl i
Start Retro in interactive mode.
.It Fl t Ar filename
Run any code and test blocks in the specified file.
.It Fl f Ar filename
Run any code blocks in the specified file.
.It Fl p Ar filename
Run the code in the specified file.
.It Fl u Ar filename
Load and use the specified image file rather than the integral one.
.It Fl r Ar filename
Load the specified image file rather than the integral one, then start Retro
in interactive mode.
.It Fl cs Ar delimiter
Use the specified delimiter to start code blocks.
.It Fl ce Ar delimiter
Use the specified delimiter to end code blocks.
.It Fl ts Ar delimiter
Use the specified delimiter to start test blocks.
.It Fl te Ar delimiter
Use the specified delimiter to end test blocks.
.It Ar filename Ar script-args
Run code blocks in a single file. Pass script-args to the code being run.
.El
.Pp
If invoked with no arguments,
.Nm
will run in interactive mode.
.Sh AUTHORS
.An Charles Childers Aq Mt crc@forthworks.com


CHANGELOG.txt

2017-10-05  Charles Childers  <crc@forthworks.com>

	* Merge branch 'roo-interface'
	* roo finished
	* saving blocks works now
	* s:evaluate no longer generates an error on empty tokens
	* implement insert mode
	* first part of support for modes finished
	* allow for exiting cleanly

2017-10-04  Charles Childers  <crc@forthworks.com>

	* remove old files; revert Makefile changes
	* limit cursor to stay in block; roo.forth now runnable directly
	* more commentary, refine initial keys a little
	* use words instead of a case construct for key handlers
	* cursor movement
	* start some work on block display code
	* begin working on roo (block editor ui)

2017-10-03  Charles Childers  <crc@forthworks.com>

	* add s:evaluate

2017-09-30  Charles Childers  <crc@forthworks.com>

	* remove some debug code from gopher.c

2017-09-29  Charles Childers  <crc@forthworks.com>

	* add gopher:get

2017-09-28  Charles Childers  <crc@forthworks.com>

	* expose files:R+ mode

2017-08-31  Charles Childers  <crc@forthworks.com>

	* update changelog

2017-08-24  Charles Childers  <crc@forthworks.com>

	* add set:reduce
	* next release will be 2017.9
	* add s:copy, document additions to s: namespace

2017-08-21  Charles Childers  <crc@forthworks.com>

	* some new constants in s: namespace
	* bump image to 2017.8
	* some tests for choose, n:negative?

2017-08-02  Charles Childers  <crc@forthworks.com>

	* tests for n:even?, n:odd?, times, n:between?
	* tests for n:min, n:max, n:negate
	* tests for n:limit, v:limit, tuck, fetch-next

2017-07-31  Charles Childers  <crc@forthworks.com>

	* add tests for gt?, lt?, gteq?, lteq?
	* add tests for c:vowel?, c:-vowel?
	* add tests for store, fetch
	* add tests for n:inc, n:dec, const
	* add tests for var, var<n>
	* add tests for v:dec, v:dec-by

2017-07-30  Charles Childers  <crc@forthworks.com>

	* add test for case, s:case
	* initial test for v:inc, v:preserve
	* add initial tests for tri, tri@, tri*
	* add tests for bi, bi@, bi*

2017-07-24  Charles Childers  <crc@forthworks.com>

	* add more tests
	* posix files: more commentary, add file:read-line, file:for-each-line
	* update CHANGELOG
	* tanu is now a filter; closes #12
	* more readability improvements to the Makefile generator
	* use the new build tool to generate a Makefile for the interfaces, extend, and embedimage

2017-07-22  Charles Childers  <crc@forthworks.com>

	* remove unused file
	* remove unused file
	* build.forth: cleaner Makefile output
	* fix listener script
	* new build tool now capable of outputting a valid Makefile
	* more work on new build tool: scan dependencies
	* build.forth: take argument from command line
	* build tool: inching closer
	* more work on build.forth
	* new example

2017-07-21  Charles Childers  <crc@forthworks.com>

	* start work on something like build.c, but in forth
	* begin early work to use new build tool for compiling sources
	* bring in new build tool (intended to help work on issue #13)
	* move documentation into doc dir
	* floating point now an io module
	* close block for floating point code
	* add a .gitattributes
	* seriously raise memory limits: 1M return stack; 4K stack depth; 48M heap
	* split rre.forth better
	* begin work to support embedding source files into rre, other interfaces

2017-07-20  Charles Childers  <crc@forthworks.com>

	* minor update to Primes example; now supports much bigger sets
	* add some more tests
	* Fix example. Closes issue #11
	* begin work on fixing argument handler. see issue #11
	* add a couple more tests

2017-07-19  Charles Childers  <crc@forthworks.com>

	* a few more tests
	* example: using VT100/ANSI escape sequences
	* fix issue w/documentation for s:with-format
	* use .forth suffix for examples, not .md
	* add note on the fenced syntax to Interfaces.md
	* new example: simple implementation of "echo"
	* add support for ~~~ code blocks as well. closes #10
	* begin work on fixing issue #10 (better support for fenced code blocks)
	* Merge pull request #9 from ratanasov/master

2017-07-19  Radoslav Atanasov  <rsrado@gmail.com>

	* Included <stdint.h> to make int32_t visible

2017-07-18  Charles Childers  <crc@forthworks.com>

	* a few more tests
	* new example
	* split legalese into separate LICENSE file
	* add 48 new tests
	* increase test coverage (all of ASCII:, some others)
	* remove some unused code
	* fix initial instance of unspec behaviour in bridge.c (see issue #8)

2017-07-17  Charles Childers  <crc@forthworks.com>

	* start bringing in tests from retro11
	* begin work on an automated testing set
	* document the file: namespace, fix stack comments in rre.forth.md
	* some additional comments on shifts in Nga documentation
	* more work on travis support
	* begin work on travis ci support
	* add sys: namespace, with argc/argv words (addresses issue #6)

2017-07-16  Charles Childers  <crc@forthworks.com>

	* correct f:to-number max value issue; fixes bug #5
	* Merge pull request #4 from ratanasov/master

2017-07-16  Radoslav Atanasov  <rsrado@gmail.com>

	* `@` and `!` added to the prefix table in `RetroForth.md`
	* `make clean` now deletes `bin/repl` too.

2017-07-12  Charles Childers  <crc@forthworks.com>

	* add n:MIN and n:MAX constant
	* fix for issue #3 (bug in n:to-string)

2017-07-11  Charles Childers  <crc@forthworks.com>

	* fix documentation issue
	* update changelog
	* documentation for FPU
	* include f: namespace by default
	* initial checkin supporting FPU (port from WIP on iOS implementation)
	* fix dump-stack

2017-07-08  Charles Childers  <crc@forthworks.com>

	* rename .s (in prep for using . as prefix for floating point)

2017-06-15  Charles Childers  <crc@forthworks.com>

	* add set:length
	* example: finding prime numbers

2017-06-14  Charles Childers  <crc@forthworks.com>

	* stdlib now includes a Retro implementation of muri assembler
	* bump image version
	* update changelog
	* merge in samples, documentation

2017-05-26  Charles Childers  <crc@forthworks.com>

	* documentation updates

interfaces/retro.py
#!/usr/bin/env python3

# Nga: a Virtual Machine
# Copyright (c) 2010 - 2018, Charles Childers
# Optimizations and process() rewrite by Greg Copeland
# -----------------------------------------------------

import os, sys, math, time, struct
from struct import pack, unpack

ip = 0
stack = [] * 128
address = []
memory = []

def rxDivMod( a, b ):
    x = abs(a)
    y = abs(b)
    q, r = divmod(x, y)
    if a < 0 and b < 0:
        r *= -1
    elif a > 0 and b < 0:
        q *= -1
    elif a < 0 and b > 0:
        r *= -1
        q *= -1
    return q, r


def findEntry(named):
    header = memory[2]
    Done = False
    while (header != 0 and not Done):
        if named == extractString(header + 3):
            Done = True
        else:
            header = memory[header]
    return header

def rxGetInput():
    return ord(sys.stdin.read(1))

def rxDisplayCharacter():
    global stack
    if stack[-1] > 0 and stack[-1] < 128:
        if stack[-1] == 8:
            sys.stdout.write(chr(stack.pop()))
            sys.stdout.write(chr(32))
            sys.stdout.write(chr(8))
        else:
            sys.stdout.write(chr(stack.pop()))
    else:
        sys.stdout.write("\033[2J\033[1;1H")
        stack.pop()
    sys.stdout.flush()

def i_no():
    pass

def i_li():
    global ip, memory, stack, address
    ip += 1
    stack.append( memory[ip] )

def i_du():
    global ip, memory, stack, address
    stack.append( stack[-1] )

def i_dr():
    global ip, memory, stack, address
    stack.pop()

def i_sw():
    global ip, memory, stack, address
    a = stack[-2]
    stack[-2] = stack[-1]
    stack[-1] = a

def i_pu():
    global ip, memory, stack, address
    address.append(stack.pop())

def i_po():
    global ip, memory, stack, address
    stack.append(address.pop())

def i_ju():
    global ip, memory, stack, address
    ip = stack.pop() - 1

def i_ca():
    global ip, memory, stack, address
    address.append(ip)
    ip = stack.pop() - 1

def i_cc():
    global ip, memory, stack, address
    target = stack.pop()
    if stack.pop() != 0:
        address.append(ip)
        ip = target - 1

def i_re():
    global ip, memory, stack, address
    ip = address.pop()

def i_eq():
    global ip, memory, stack, address
    a = stack.pop()
    b = stack.pop()
    if b == a:
        stack.append(-1)
    else:
        stack.append(0)

def i_ne():
    global ip, memory, stack, address
    a = stack.pop()
    b = stack.pop()
    if b != a:
        stack.append(-1)
    else:
        stack.append(0)

def i_lt():
    global ip, memory, stack, address
    a = stack.pop()
    b = stack.pop()
    if b < a:
        stack.append(-1)
    else:
        stack.append(0)

def i_gt():
    global ip, memory, stack, address
    a = stack.pop()
    b = stack.pop()
    if b > a:
        stack.append(-1)
    else:
        stack.append(0)

def i_fe():
    global ip, memory, stack, address
    stack[-1] = memory[stack[-1]]

def i_st():
    global ip, memory, stack, address
    mi = stack.pop()
    memory[mi] = stack.pop()

def i_ad():
    global ip, memory, stack, address
    t = stack.pop()
    stack[ -1 ] += t
    stack[-1] = unpack('=l', pack('=L', stack[-1] & 0xffffffff))[0]

def i_su():
    global ip, memory, stack, address
    t = stack.pop()
    stack[-1] -= t
    stack[-1] = unpack('=l', pack('=L', stack[-1] & 0xffffffff))[0]

def i_mu():
    global ip, memory, stack, address
    t = stack.pop()
    stack[-1] *= t
    stack[-1] = unpack('=l', pack('=L', stack[-1] & 0xffffffff))[0]

def i_di():
    global ip, memory, stack, address
    a = stack[-1]
    b = stack[-2]
    stack[-1], stack[-2] = rxDivMod( b, a )
    stack[-1] = unpack('=l', pack('=L', stack[-1] & 0xffffffff))[0]
    stack[-2] = unpack('=l', pack('=L', stack[-2] & 0xffffffff))[0]

def i_an():
    global ip, memory, stack, address
    t = stack.pop()
    stack[-1] &= t

def i_or():
    global ip, memory, stack, address
    t = stack.pop()
    stack[-1] |= t

def i_xo():
    global ip, memory, stack, address
    t = stack.pop()
    stack[-1] ^= t

def i_sh():
    global ip, memory, stack, address
    t = stack.pop()
    stack[-1] <<= t
    t = stack.pop()
    stack[-1] >>= t

def i_zr():
    global ip, memory, stack, address
    if stack[-1] == 0:
        stack.pop()
        ip = address.pop()

def i_en():
    global ip, memory, stack, address
    ip = 9000000

def i_ie():
    stack.push(1)

def i_iq():
    stack.pop()
    stack.push(0)
    stack.push(0)

def i_ii():
    stack.pop()
    rxDisplayCharacter()

instructions = [i_no, i_li, i_du, i_dr, i_sw, i_pu, i_po, i_ju, i_ca, i_cc, i_re, i_eq, i_ne, i_lt, i_gt, i_fe, i_st, i_ad, i_su, i_mu, i_di, i_an, i_or, i_xo, i_sh, i_zr, i_en, i_ie, i_iq, i_ii]

def validateOpcode(opcode):
    I0 = opcode & 0xFF
    I1 = (opcode >> 8) & 0xFF
    I2 = (opcode >> 16) & 0xFF
    I3 = (opcode >> 24) & 0xFF
    if (I0 >= 0 and I0 <= 29) and (I1 >= 0 and I1 <= 29) and (I2 >= 0 and I2 <= 29) and (I3 >= 0 and I3 <= 29):
        return True
    else:
        return False

def extractString( at ):
    i = at
    s = ''
    while (memory[i] != 0):
        s = s + chr(memory[i])
        i = i + 1
    return s

def injectString( s, to ):
    global memory
    i = to
    for c in s:
        memory[i] = ord(c)
        i = i + 1
    memory[i] = 0

def execute(word, output = 'console'):
    global ip, memory, stack, address
    ip = word
    address.append(0)
    notfound = memory[findEntry('err:notfound') + 1]
    while ip < 100000 and len(address) > 0:
        if ip == notfound:
            print('ERROR: word not found!')
        opcode = memory[ip]
        if validateOpcode(opcode):
            I0 = opcode & 0xFF
            I1 = (opcode >> 8) & 0xFF
            I2 = (opcode >> 16) & 0xFF
            I3 = (opcode >> 24) & 0xFF
            if I0 != 0: instructions[I0]()
            if I1 != 0: instructions[I1]()
            if I2 != 0: instructions[I2]()
            if I3 != 0: instructions[I3]()
        else:
            print('Invalid Bytecode', opcode, ip)
            ip = 2000000
        ip = ip + 1
    return

def load_image():
    global memory
    cells = int(os.path.getsize('ngaImage') / 4)
    f = open( 'ngaImage', 'rb' )
    memory = list(struct.unpack( cells * 'i', f.read() ))
    f.close()
    remaining = 1000000 - cells
    memory.extend( [0] * remaining )

def run():
    Done = False
    Interpreter = memory[findEntry('interpret') + 1]
    while not Done:
        Line = input('\nOk> ')
        if Line == 'bye': Done = True
        else:
            for Token in Line.split(' '):
                injectString(Token, 1025)
                stack.append(1025)
                execute(Interpreter)

if __name__ == "__main__":
    load_image()
    run()

